rclpy/rclpy/time_source.py

```diff
@@ -0,0 +1,96 @@
+# Copyright 2018 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import builtin_interfaces.msg
+from rclpy.clock import Clock
+from rclpy.clock import ClockType
+from rclpy.clock import ROSClock
+from rclpy.time import Time
+
+CLOCK_TOPIC = '/clock'
+
+
+class TimeSource:
+
+    def __init__(self, *, node=None):
+        self._clock_sub = None
+        self._node = None
+        self._associated_clocks = []
+        self._last_time_set = None
+        self._ros_time_is_active = False
+        if node is not None:
+            self.attach_node(node)
+
+    @property
+    def ros_time_is_active(self):
+        return self._ros_time_is_active
+
+    @ros_time_is_active.setter
+    def ros_time_is_active(self, enabled):
+        self._ros_time_is_active = enabled
+        for clock in self._associated_clocks:
+            clock._set_ros_time_is_active(enabled)
+        if enabled:
+            self._subscribe_to_clock_topic()
+
+    def _subscribe_to_clock_topic(self):
+        if self._clock_sub is None and self._node is not None:
+            self._clock_sub = self._node.create_subscription(
+                builtin_interfaces.msg.Time,
+                CLOCK_TOPIC,
+                self.clock_callback
+            )
+
+    def attach_node(self, node):
+        from rclpy.node import Node
+        if not isinstance(node, Node):
+            raise TypeError('Node must be of type rclpy.node.Node')
+        # Remove an existing node.
+        if self._node is not None:
+            self.detach_node()
+        self._node = node
+        if self.ros_time_is_active:
+            self._subscribe_to_clock_topic()
+
+    def detach_node(self):
+        # Remove the subscription to the clock topic.
+        if self._clock_sub is not None:
+            if self._node is None:
+                raise RuntimeError('Unable to destroy previously created clock subscription')
+            self._node.destroy_subscription(self._clock_sub)
+        self._clock_sub = None
+        self._node = None
+
+    def attach_clock(self, clock):
+        if not isinstance(clock, Clock):
+            raise TypeError('Clock must be of type rclpy.clock.Clock')
+        if not clock.clock_type == ClockType.ROS_TIME:
+            raise ValueError('Only clocks with type ROS_TIME can be attached.')
+        # "Cast" to ROSClock subclass so ROS time methods can be used.
+        clock.__class__ = ROSClock
```

Comment:
Not asking to change anything, just sharing because I think it's interesting. This led me discover that `__new__` could be used to return a `ROSClock` instance when passed a ROS clock type.

<details>
<summary>Details</summary>

```python

class Clock:

    def __new__(cls, *, clock_type="SYSTEM_CLOCK"):
        self = None
        if clock_type == "ROS_CLOCK":
            self = super().__new__(ROSClock)
        else:
            self = super().__new__(cls)
        self._clock_type = clock_type
        # create self._clock_handle here too
        return self

    # Can't have an __init__ because it would get passed clock_type argument
    # and ROSCLOCK doesn't wan't one
    # def __init__(self):
    #   pass

    def all_clocks_have_this(self):
        print("all clocks have this")


class ROSClock(Clock):

    # Can't have an __init__ because it would get passed clock_type argument
    # and ROSCLOCK doesn't wan't one
    # def __init__(self):
    #   pass

    def only_ros_clocks_have_this(self):
        print("Only ros clocks have this")

c1 = Clock()
c2 = Clock(clock_type="STEADY_CLOCK")
c3 = Clock(clock_type="ROS_CLOCK")

print(repr(c1))
print(repr(c2))
print(repr(c3))

c1.all_clocks_have_this()
c2.all_clocks_have_this()
c3.all_clocks_have_this()
c3.only_ros_clocks_have_this()
```
</details>

Reply:
I think this approach might be better! Because there's no extra step for users to get access to ROS clock methods if they have created a ROS clock with `Clock(clock_type=ClockType.ROS_TIME)`, they will always get a ROSClock back.

I don't immediately see any disadvantages of this (it's ok in this instance that the Clock class 'knows about' the ROSClock class), so I plan to incorporate it, thank you for the suggestion

Reply:
Done in 9992b1f224570c9f18f5ed26156e457602a267fb + cf38ae765503bff335b624c890ca5d0a7d30e227, I think it's better to use now: happy you pointed this out!